agcaballero/walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -33,68 +37,126 @@
    */ 
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
     Collection<String> attendees = request.getAttendees();
-    ArrayList<TimeRange> times = new ArrayList<TimeRange>();
-     // minutes of day is + 1 to account for first and last minute of day
-    boolean[] minutes = new boolean[MINUTES_IN_DAY + 1];
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
 
-    for(int i = 0; i < minutes.length; i++) {
-      minutes[i] = true;
-    }
+    // minutes of day is + 1 to account for first and last minute of day
+    Availability[] minutes = new Availability[MINUTES_IN_DAY + 1];
+    // set all spots to default value of all available
+    Arrays.fill(minutes, Availability.ALL_AVAILABLE);
 
-    for(Event event: events) {
-      // if there's an overlap in attendees block off those times
-      if(attendeeOverlap(event.getAttendees(), attendees)) {
+    for (Event event: events) {
+      Availability status = Availability.ALL_AVAILABLE;
+      
+      if (! Collections.disjoint(event.getAttendees(), attendees)) {
+        // if there's overlap in mandatory attendees, time is unavailable
+        status = Availability.UNAVAILABLE;
+      } else if (! Collections.disjoint(event.getAttendees(), optionalAttendees)) {
+        // if there's an overlap in only the optional attendees, time is available for only the mandatory employees
+        status = Availability.MANDATORY_AVAILABLE;
+      }
+
+      if (status != Availability.ALL_AVAILABLE) {
```

Comment:
This block of code will need to be updated if the enum gets additional values.

Maybe it would be safer (less likely to forget to update this block) to use an exhaustive switch, so that if a new enum value is added, it will also need to be addressed here:

```
switch (status) {
  case  ALL_AVAILABLE:
    continue;
  case UNAVAILABLE:
    Arrays.fill(minutes, range.start(), range.end(), UNAVAILABLE);
    break;
  case MANDATORY_AVAILABLE:
    ...
    break;
}
```

Reply:
That makes sense. I will add this.

Reply:
Done